src/hierarchies/nnw_hierarchy.h

```diff
@@ -28,94 +28,64 @@
 //! posterior distribution are available in closed form and Neal's algorithm 2
 //! may be used with it.
 
-class NNWHierarchy : public BaseHierarchy {
- public:
-  struct State {
-    Eigen::VectorXd mean;
-    Eigen::MatrixXd prec;
-  };
-  struct Hyperparams {
-    Eigen::VectorXd mean;
-    double var_scaling;
-    double deg_free;
-    Eigen::MatrixXd scale;
-    Eigen::MatrixXd scale_inv;
-  };
+namespace NNW {
+struct State {
+  Eigen::VectorXd mean;
+  Eigen::MatrixXd prec;
+  Eigen::MatrixXd prec_chol;
+  double prec_logdet;
+};
+struct Hyperparams {
+  Eigen::VectorXd mean;
+  double var_scaling;
+  double deg_free;
+  Eigen::MatrixXd scale;
+  Eigen::MatrixXd scale_inv;
+};
+}  // namespace NNW
 
+class NNWHierarchy
+    : public BaseHierarchy<NNWHierarchy, NNW::State, NNW::Hyperparams,
+                           bayesmix::NNWPrior> {
  protected:
   unsigned int dim;
   Eigen::VectorXd data_sum;
   Eigen::MatrixXd data_sum_squares;
-  // STATE
-  State state;
-  // HYPERPARAMETERS
-  std::shared_ptr<Hyperparams> hypers;
-  Hyperparams posterior_hypers;
-
-  // UTILITIES FOR LIKELIHOOD COMPUTATION
-  //! Lower factor of the Cholesky decomposition of state.prec
-  Eigen::MatrixXd prec_chol;
-  //! Determinant of state.prec in logarithmic scale
-  double prec_logdet;
 
   // AUXILIARY TOOLS
   //! Special setter for prec and its utilities
-  void set_prec_and_utilities(const Eigen::MatrixXd &prec_);
-  //!
-  void clear_data() override;
-  //!
-  void update_summary_statistics(const Eigen::VectorXd &datum,
-                                 const Eigen::VectorXd &covariate,
-                                 bool add) override;
-  //!
-  void save_posterior_hypers() override;
-  //!
-  void initialize_hypers() override;
-  //! Returns updated values of the prior hyperparameters via their posterior
-  Hyperparams normal_wishart_update() const;
-  //!
-  std::shared_ptr<bayesmix::NNWPrior> cast_prior() {
-    return std::dynamic_pointer_cast<bayesmix::NNWPrior>(prior);
-  }
-  //!
-  void create_empty_prior() override { prior.reset(new bayesmix::NNWPrior); }
+  void set_prec_and_utilities(const Eigen::MatrixXd &prec_, NNW::State *out);
```

Comment:
If this now receives a proto, maybe we should change its name. Maybe `set_prec_to_state()`? Since the utilities now also belong to the state, we don't have to specify it in the name. Just like we don't specify updating the inverse of matrices.

Reply:
What about `write_prec_to_state` ?

Reply:
Sure.